### ISSUE TITLE
BATS tests don't show up in CircleCI as tests

### ISSUE THREAD

**Issue by Peter-Darton-i2** [CircleCI-Public/BATS-orb#4]
**Orb Version**
1.0.0

**Describe the bug**
CircleCI expects that test jobs report their tests to CircleCI so that the tests can be included in the CircleCI test report.
BATS is a test framework that runs tests.
...but the `bats/run` job doesn't give the tests to CircleCI, which makes debugging difficult.

**To Reproduce**
1. Follow the instructions in https://circleci.com/docs/testing-orbs/#unit-testing
2. Create a simple BATS file that runs some tests; make some of the pass, some fail, have all of them output some text or other.
3. Run the build in CircleCI
4. Go to the job `bats/run`
5. See that there are no test results in the CircleCI UI - as far as it's concerned, it ran a command and the command failed, but it provides no further details.

**Expected behavior**
Each BATS test case should appear in the CircleCI UI as its own test case.
The user should be able to use the CircleCI WebUI's "TESTS" tab to see whether each test passed or failed _without_ having to inspect the console output of the bats/run job.
...and should be able to inspect the times each took, stdout, stderr etc just as for any other "proper" unit-test framework
...and, ideally, should be able to split the tests to support parallel execution.

**Additional context**

This orb is called out in the [CircleCI documentation](https://circleci.com/docs/testing-orbs/#unit-testing) as being the technology of choice for testing bash scripts used in CircleCI orbs, which leads to an expectation that this orb would at least support "normal" test functionality with CircleCI.
The lack of any `store_test_results` functionality in this orb was a surprise - tests _should_ appear in the tests tab.

TL;DR: `bats/run` should call `store_test_results` and upload the results.

**Comment by Peter-Darton-i2** [comment]
I've done some further reading; it looks like CircleCI doesn't understand the TAP unit-test format so it can't ingest BATS' output directly.
...but the existing CircleCI documentation for testing [Ava for Node.js](https://circleci.com/docs/collect-test-data/#ava-for-node) instructs users to convert TAP output (from ava in this case) into xunit output using [tap-xunit](https://github.com/aghassemi/tap-xunit)

i.e. there's prior-art for handling a TAP-compliant test framework in CircleCI.

e.g. if `bats/run` were modified to ensure that `tap-xunit` was installed, it could then pipe bats' output through e.g. `| tee /tmp/bats.tap` and then `cat /tmp/bats.tap | tap-xunit > /tmp/bats.xml` before doing a `store_test_results` with `path: /tmp/bats.xml`.
Perhaps such modification could be made conditional on a parameter being passed in, e.g. one that's then passed to `tap-xunit`'s `--package` argument ... although it would be better if the path to the .bats file(s) being run was incorporated into the package name too.

**Comment by Peter-Darton-i2** [comment]
I've done a proof-of-concept.
```
commands:
  install_tap2junit:
    description: >
      Installs a utility that can convert TAP format unit-test data
      into a format that CircleCI can understand.
    steps:
      - run:
          name: Install tap2junit
          command: |
            set -e
            set -o pipefail
            checkRequirements() {
              if ! which apt-get > /dev/null; then
                echo "apt-get is required to install tap2junit"
                exit 1
              fi
              if [[ $EUID == 0 ]]; then export SUDO=""; else export SUDO="sudo"; fi
            }
            doInstall() {
              echo '+' $SUDO apt-get install libtap-formatter-junit-perl
              $SUDO apt-get install libtap-formatter-junit-perl
            }
            checkInstall() {
              if ! which tap2junit > /dev/null; then
                echo "tap2junit has failed to install."
                return 1
              fi
              echo
              echo "tap2junit installed"
              echo
            }
            runInstall() {
              checkRequirements
              doInstall
              checkInstall
            }
            if ! which tap2junit > /dev/null 2>&1; then
              runInstall
            fi
```
and then, in the job, instead of doing
```
      - bats/install
      - steps: <<parameters.setup-steps>>
      - run:
          name: Execute BATS tests
          command: bats <<parameters.path>>
```
we do:
```
      - bats/install
      - install_tap2junit
      - steps: <<parameters.setup-steps>>
      - run:
          name: Execute BATS tests
          command: |
            set -e
            set -o pipefail
            mkdir -p /tmp/bats-tap-data
            if bats --tap <<parameters.path>> | tee /tmp/bats-tap-data/bats.tap; then
              tap2junit --verbose /tmp/bats-tap-data/bats.tap
            else
              return_code=$?
              tap2junit --verbose /tmp/bats-tap-data/bats.tap
              exit ${return_code}
            fi
      - store_test_results:
          name: Upload converted TAP data to CircleCI
          path: /tmp/bats-tap-data/bats.tap.xml
```

This results in the (perl-based) tap2junit utility being installed (if it isn't already there), BATS being told to generate TAP-format data, which is captured in a file, which is turned into JUnit XML _without_ suppressing the command's return status ... and the JUnit XML data is uploaded to CircleCI.

Note: I did try out the `tap-xunit` utility (as suggested in https://circleci.com/docs/collect-test-data/#ava-for-node) but tap2junit was better (in case of a test failure, tap2junit makes the full test output visible in CircleCI whereas tap-xunit just gave a pass/fail status).

**Comment by KyleTryon** [comment]
Hey @Peter-Darton-i2 sorry for the wait.

Sorry we didn't implement this here yet either, we actually already do exactly this inside the orb-tools-orb for the "review check" job. I'm going to get this added here 👍

**Comment by Peter-Darton-i2** [comment]
Funnily enough, one of your support folks said I should submit a PR and you'd look at it, and so I was just putting one together.
I've pushed what I've got so far - you're welcome to grab it.

**Comment by KyleTryon** [comment]
Changes included via #10 should hopefully resolve this request. Please let us know. The release will be out shortly and will be published as v1.1.0
